src/coreclr/jit/morphblock.cpp

```diff
@@ -1358,8 +1358,13 @@ GenTree* MorphCopyBlockHelper::CopyFieldByField()
                 }
                 else
                 {
-                    GenTree* fldAddr = grabAddr(srcFieldOffset);
-                    dstFldStore      = m_comp->gtNewStoreIndNode(srcType, fldAddr, srcFld);
+                    GenTree*     fldAddr    = grabAddr(srcFieldOffset);
+                    GenTreeFlags indirFlags = GTF_EMPTY;
+                    if (m_store->OperIs(GT_STORE_BLK, GT_STOREIND))
+                    {
+                        indirFlags = m_store->gtFlags & (GTF_IND_TGT_NOT_HEAP | GTF_IND_TGT_HEAP);
+                    }
+                    dstFldStore = m_comp->gtNewStoreIndNode(srcType, fldAddr, srcFld, indirFlags);
```

Comment:
Here we were losing `GTF_IND_TGT*` flags when we promote GT_STORE_BLK to field-wise GT_STOREIND